src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs

```diff
@@ -3,91 +3,317 @@
+
+    public static async ValueTask WriteDataAsync(
+        PipeWriter pipeWriter,
+        AssetPath assetPath,
+        ReadOnlyMemory<Checksum> checksums,
+        SolutionAssetStorage.Scope scope,
+        ISerializerService serializer,
+        CancellationToken cancellationToken)
+    {
+        // Create a channel to communicate between the searching and writing tasks.  This allows the searching task to
+        // find items, add them to the channel synchronously, and immediately continue searching for more items.
+        // Concurrently, the writing task can read from the channel and write the items to the pipe-writer.
+        var channel = Channel.CreateUnbounded<(Checksum checksum, object asset)>(s_channelOptions);
+
+        // When cancellation happens, attempt to close the channel.  That will unblock the task writing the assets
+        // to the pipe. Capture-free version is only available on netcore unfortunately.
+#if NET
+        using var _ = cancellationToken.Register(
+            static (obj, cancellationToken) => ((ChecksumChannel)obj!).Writer.TryComplete(new OperationCanceledException(cancellationToken)),
+            state: channel);
+#else
+        using var _ = cancellationToken.Register(
+            () => channel.Writer.TryComplete(new OperationCanceledException(cancellationToken)));
+#endif
+
+        // Spin up a task to go search for all the requested checksums, adding results to the channel.
+        var findAssetsTask = FindAllAssetsAsync(assetPath, checksums, scope, channel, cancellationToken);
+
+        // Spin up a task to read from the channel and write out the assets to the pipe-writer.
+        var writeAssetsTask = WriteAllAssetsToPipeAsync(
+            pipeWriter, channel, checksums.Length, scope, serializer, cancellationToken);
+
+        // Wait for both the searching and writing tasks to finish.
+        await Task.WhenAll(findAssetsTask, writeAssetsTask).ConfigureAwait(false);
+
+        return;
+
+        static async Task FindAllAssetsAsync(
+            AssetPath assetPath,
             ReadOnlyMemory<Checksum> checksums,
+            SolutionAssetStorage.Scope scope,
+            ChecksumChannel channel,
             CancellationToken cancellationToken)
         {
-            using var writer = new ObjectWriter(stream, leaveOpen: true, cancellationToken);
+            await Task.Yield();
 
-            // This information is not actually needed on the receiving end.  However, we still send it so that the
-            // receiver can assert that both sides are talking about the same solution snapshot and no weird invariant
-            // breaks have occurred.
-            solutionChecksum.WriteTo(writer);
+            Exception? exception = null;
+            try
+            {
+                await scope.FindAssetsAsync(
+                    assetPath,
+                    checksums,
+                    static (checksum, asset, channel) => channel.Writer.TryWrite((checksum, asset)),
```

Comment:
>TryWrite [](http://example.com/codeflow?start=72&length=8)

What does it mean if this returns false?

Reply:
false only happens if the writer has been closed.  which only happens if we've been cancelled.  so it's fine for this to be ignored.  i'll doc.